base/tps/src/org/dogtagpki/server/tps/rest/ActivityService.java

```diff
@@ -184,24 +254,49 @@ protected void retrieveActivitiesWithoutVLV(
         // count the total entries
         for (; activities.hasNext(); i++) activities.next();
         response.setTotal(i);
+	*/
     }
 
     @Override
     public Response getActivity(String activityID) {
 
+        String method = "ActivityService.getActivity: ";
+	String msg = "";
         if (activityID == null) throw new BadRequestException("Activity ID is null.");
 
-        CMS.debug("ActivityService.getActivity(\"" + activityID + "\")");
+        CMS.debug(method + "(\"" + activityID + "\")");
 
         try {
+            List<String> authorizedProfiles = getAuthorizedProfiles();
+
             TPSSubsystem subsystem = (TPSSubsystem)CMS.getSubsystem(TPSSubsystem.ID);
             ActivityDatabase database = subsystem.getActivityDatabase();
+	    ActivityRecord record = database.getRecord(activityID);
+	    String type = record.getType();
 
-            return createOKResponse(createActivityData(database.getRecord(activityID)));
+            if ((type != null) && !authorizedProfiles.contains(UserResource.ALL_PROFILES) && !authorizedProfiles.contains(type)) {
+                msg = "type not authorized: " + type;
+                CMS.debug(method + msg);
+                throw new PKIException(msg);
+	    }
+            return createOKResponse(createActivityData(record));
 
         } catch (Exception e) {
             CMS.debug(e);
             throw new PKIException(e.getMessage());
         }
     }
+
+    /*
+     * returns a list of TPS profiles allowed for the current user
+     */
+    List<String> getAuthorizedProfiles()
+           throws Exception {
+        String method = "TokenService.getAuthorizedProfiles: ";
+        String userID = servletRequest.getUserPrincipal().getName();
+        CMS.debug(method + "principal name: " + userID);
+        IUGSubsystem userGroupManager = (IUGSubsystem) CMS.getSubsystem(CMS.SUBSYSTEM_UG);
+        IUser user = userGroupManager.getUser(userID);
+        return user.getTpsProfiles();
```

Comment:
I think it should be possible to downcast the user principal into `PKIPrincipal` and get the `IUser` object like this:
```
PKIPrincipal pkiPrincipal = (PKIPrincipal) servletRequest.getUserPrincipal();
IUser user = pkiPrincipal.getUser();
return user.getTpsProfiles();
```
So it's not necessary to search the user again. The `PKIPrincipal` is created earlier by the authentication realm:
https://github.com/dogtagpki/pki/blob/master/base/server/src/main/java/com/netscape/cms/realm/PKIRealm.java#L198

Reply:
good idea!  Will try it out. Thanks!